libs/sdk-ui/src/base/react/placeholders/docs.md

```diff
@@ -0,0 +1,385 @@
+
+Placeholders are built on top of React context and hooks.
+
+### Getting started
+
+###### 1/ Wrap your application in PlaceholdersProvider
+
+```
+const Root = () =>  (
+    <PlaceholdersProvider>
+        <App />
+    </PlaceholdersProvider>
+);
+```
+
+###### 2/ Create your first placeholder
+
+```
+const primaryMeasurePlaceholder = newPlaceholder<IMeasure>();
+```
+
+###### 3/ Use your first placeholder
+
+```
+<BarChart
+    measures={[primaryMeasurePlaceholder]}
+/>
+```
+
+###### 4/ Change value of your first placeholder
+
+```
+const PrimaryMeasureSelect = () => {
+    const [primaryMeasure, setPrimaryMeasure] = primaryMeasurePlaceholder.use();
+
+    return (
+        <>
+            Active primary measure identifier: {
+                primaryMeasure
+                    ? measureIdentifier(primaryMeasure)
+                    : "No primary measure"
+            }
+            <button onClick={() => setPrimaryMeasure(Md.Revenue)}>Revenue</button>
+        </>
+    );
+};
+```
+
+### Common placeholders
+
+Common placeholders may hold one or multiple values (attributes/measures/filters/sorts...). Note that you should not set other placeholders as the placeholder value - to combine multiple placeholders, use composed placeholder instead. You can create common placeholder by calling `newPlaceholder` method from `@gooddata/sdk-ui`.
+
+###### Create common placeholder
+
+```
+// Create common placeholder that may hold single measure.
+const measurePlaceholder = newPlaceholder<IMeasure>();
+
+// Create common placeholder that may hold array of attributes.
+const attributesPlaceholder = newPlaceholder<IAttribute[]>();
+```
+
+###### Set default value of the common placeholder
+
+```
+// Create common placeholder with default value.
+// Note that in this case, type of the placeholder is derived from the default value,
+// so you don't have to explicitly provide it.
+const measurePlaceholder = newPlaceholder(Md.Revenue);
+
+// In this case, you have to provide the typing as [] would be infered as never[].
+const measuresPlaceholder = newPlaceholder<IMeasure[]>([]);
+```
+
+###### Provide common placeholder to the visualization
+
+```
+// You can provide placeholder instead of the concrete value to any visualization.
+// Placeholder will be replaced by its set value or the default value.
+<BarChart
+    measures={[measurePlaceholder]}
+/>
+
+// You can provide also placeholders that may hold array of values - it will be flattened.
+<BarChart
+    measures={[measuresPlaceholder]}
+/>
+```
+
+###### Set validation and other options of the common placeholder
+
+```
+// Set custom validation or id of the placeholder.
+const measurePlaceholder = newPlaceholder(Md.Revenue, {
+
+    // For the debugging, it can be useful to speficy readable id.
+    id: "primaryMeasure",
+
+    // Specify validation of the placeholder value.
+    validate: (measure) => {
+        // assert something
```

Comment:
We should be explicit about the fact the validation function should throw in case of invalid input.